### PR TITLE
[FC] Eagerly fetch chunks in the cow store

### PR DESCRIPTION
**Pull request by maggie-lou**
If we access a snapshot chunk, use a background queue to eagerly fetch the next X contiguous chunks, anticipating that they will also be accessed.

**Some performance results:**
The performance tests build the buildbuddy repo from a snapshot that is cached in the remote cache, but nothing is cached locally in filecache. Every testing scenario was run 6 times. 

Baseline: (45.8s - 50.6s)
Best config - 8 concurrent threads, max 1000 fetches per sec: (24.9s - 27.2s)

**Some additional sample tests for context:**

4 concurrent threads, max 1000 fetches per sec: (25.7s - 27.6s)
4 concurrent threads, max 500 fetches per sec: (28.5 - 29.5s)
8 concurrent threads, max 100 fetches per sec: (37.5s - 40.2)
-  Additional threads not that helpful, because rate limiter is so low
4 concurrent threads, max 100 fetches per sec: (38.1 - 42.3)
4 concurrent threads, max 1000 per sec, 10 contiguous chunks: (27.5 - 30)
- All other tests are with 4 contiguous chunks
- Fetching more chunks each time leads to a slowdown in performance

### CODE REVIEW

**Comment by maggie-lou** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5076** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/5076" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/5076?utm_source=stack-comment).

**Review by bduffany — COMMENTED** [review]
This implementation looks great.

Before submitting, I think it'd be good to do some performance testing, possibly with different tuning values (e.g. N=0 vs N=4 / concurrency=4 vs concurrency=8, etc.). Otherwise we're a bit blind as to whether the eager fetching provides a net benefit, or how things will change if we try to adjust these tuning values.

File: enterprise/server/remote_execution/copy_on_write/copy_on_write.go [hunk 1]
```diff
@@ -473,6 +516,73 @@ func (s *COWStore) initDirtyChunk(offset int64, size int64) (ogChunk *Mmap, newC
...
+func (s *COWStore) fetchChunk(offset int64) error {
+	s.mu.RLock()
+	c := s.chunks[offset]
+	s.mu.RUnlock()
+
+	// Skip holes
+	if c == nil {
+		return nil
+	}
+
```
Comment (bduffany) [inline enterprise/server/remote_execution/copy_on_write/copy_on_write.go:578]: nit: maybe move the rateLimiter.Wait() after this line?

**Comment by bduffany** [comment]
As a rough performance test, it might be useful to try running `enterprise/tools/vmstart` now that it supports remote snapshots. Some behavior I noticed was that the first LoadSnapshot took a long time (~30s), then each subsequent one got faster

File: enterprise/server/remote_execution/copy_on_write/copy_on_write.go (hunk 1 above)
Comment (maggie-lou) [reply enterprise/server/remote_execution/copy_on_write/copy_on_write.go:578]: I'm wondering if the rate limiter could also help protect over-locking on `chunks`. I'm just going to leave it as is for now - probably wouldn't be a big difference either way

**Comment by maggie-lou** [comment]
> This implementation looks great.
> 
> Before submitting, I think it'd be good to do some performance testing, possibly with different tuning values (e.g. N=0 vs N=4 / concurrency=4 vs concurrency=8, etc.). Otherwise we're a bit blind as to whether the eager fetching provides a net benefit, or how things will change if we try to adjust these tuning values.

I added some performance test results to the pr body and updated the config to match the results. Also made them configurable by flag

**Review by bduffany — APPROVED** [review]

**Pull request merged**